### PR TITLE
Feature/sync with case clash names

### PR DESCRIPTION
**Pull request by mgallien**
https://github.com/nextcloud/desktop/issues/5099

### CODE REVIEW

File: src/libsync/discovery.cpp [hunk 1]
```diff
@@ -1897,4 +1904,61 @@ void ProcessDirectoryJob::setupDbPinStateActions(SyncJournalFileRecord &record)
...
+}
+
+bool ProcessDirectoryJob::checkCaseClash(const QString &newFileName, const QString &existingFileName) const
+{
+#ifdef Q_OS_MAC
+    // Need to normalize to composited form because of QTBUG-39622/QTBUG-55896
+    const auto newFileNameComposited = newFileName.normalized(QString::NormalizationForm_C);
+    const auto existingFileNameComposited = existingFileName.normalized(QString::NormalizationForm_C);
+
+    if (newFileNameComposited.endsWith(existingFileNameComposited, Qt::CaseInsensitive)) {
```
Comment (claucambra) [inline src/libsync/discovery.cpp:1945]: Wouldn't `QString::compare` be more apt here?

File: src/libsync/discovery.cpp [hunk 2]
```diff
@@ -1897,4 +1904,61 @@ void ProcessDirectoryJob::setupDbPinStateActions(SyncJournalFileRecord &record)
         record._type = ItemTypeVirtualFileDownload;
 }
 
+void ProcessDirectoryJob::processCaseClashConflicts(const PathTuple &path,
+                                                    LocalInfo &localEntry,
+                                                    [[maybe_unused]] const RemoteInfo &serverEntry,
+                                                    const SyncJournalFileRecord &dbEntry,
+                                                    const std::map<QString, Entries> &entries) const
+{
+    for(const auto &entry : entries) {
```
Comment (claucambra) [inline src/libsync/discovery.cpp:1913]: Using a structured binding here instead of `entry` would make this easier to read IMO:

```suggestion
    for(const auto &[key, entry] : entries) {
```

would save from having to call `entry.second` repeatedly (and also better match the fact that `entry.second` is the actual entry

File: src/libsync/propagatedownload.cpp [hunk 3]
```diff
@@ -1121,14 +1121,7 @@ void PropagateDownloadFile::contentChecksumComputed(const QByteArray &checksumTy
...
     ASSERT(!_tmpFile.isOpen());
-    QString fn = propagator()->fullLocalPath(_item->_file);
-
-    // In case of file name clash, report an error
-    // This can happen if another parallel download saved a clashing file.
-    if (propagator()->localFileNameClash(_item->_file)) {
-        done(SyncFileItem::FileNameClash, tr("File %1 cannot be saved because of a local file name clash!").arg(QDir::toNativeSeparators(_item->_file)));
-        return;
-    }
+    auto filename = propagator()->fullLocalPath(_item->_file);
```
Comment (claucambra) [inline src/libsync/propagatedownload.cpp:1124]:
```suggestion
    const auto filename = propagator()->fullLocalPath(_item->_file);
```

**Review by claucambra — CHANGES_REQUESTED** [review]
Just some small nitpicks :)

**Comment by claucambra** [comment]
The sync engine test has crashed in the Windows pipeline, unsure if this is related to the changes

**Comment by tobiasKaminsky** [comment]
image

--> there should be a button "resolve conflict", so that it is better visible to the user how to fix this.

File: CMakeLists.txt [hunk 4]
```diff
@@ -168,6 +168,9 @@ option(BUILD_LIBRARIES_ONLY "BUILD_LIBRARIES_ONLY" OFF)
 # build the GUI component, when disabled only nextcloudcmd is built
 option(BUILD_GUI "BUILD_GUI" ON)
 
+# build the tests
+option(BUILD_TESTING "BUILD_TESTING" ON)
```
Comment (allexzander) [inline CMakeLists.txt:172]: why is this needed?

**Comment by allexzander** [comment]
@matthieugallien Could you add a short video of how this feature works?

File: src/common/filesystembase.cpp [hunk 5]
```diff
@@ -318,6 +318,7 @@ bool FileSystem::fileExists(const QString &filename, const QFileInfo &fileInfo)
     // fileInfo is re-used.
     if (fileInfo.filePath() != filename) {
         QFileInfo myFI(filename);
+        Q_ASSERT(myFI.filePath() == filename);
```
Comment (allexzander) [inline src/common/filesystembase.cpp:321]: you might want to add a log here as the assert won't work in Release build

File: src/csync/csync_exclude.cpp [hunk 6]
```diff
@@ -206,9 +206,13 @@ static CSYNC_EXCLUDE_TYPE _csync_excluded_common(const QString &path, bool exclu
     }
 
 
-    if (excludeConflictFiles && OCC::Utility::isConflictFile(path)) {
+    if (excludeConflictFiles && OCC::Utility::isConflictFile(path) && !OCC::Utility::isCaseClashConflictFile(path)) {
```
Comment (allexzander) [inline src/csync/csync_exclude.cpp:209]: you could make this condition more simple, like:
```
if (excludeConflictFiles) {
  if (OCC::Utility::isCaseClashConflictFile(path)) {
    return CSYNC_FILE_EXCLUDE_CASE_CLASH_CONFLICT;
   }
   if (OCC::Utility::isConflictFile(path)) {
     return CSYNC_FILE_EXCLUDE_CONFLICT;
   }
}
```

File: src/gui/caseclashfilenamedialog.h [hunk 7]
```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) by Felix Weilbach <felix.weilbach@nextcloud.com>
```
Comment (allexzander) [inline src/gui/caseclashfilenamedialog.h:2]: was this done by Felix?

File: src/libsync/caseclashconflictsolver.h [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    void onRemoteSourceFileDoesNotExist();
+
+    void onMoveJobFinished();
+
+private:
+    [[nodiscard]] QString remoteNewFilename() const;
+
+    [[nodiscard]] QString remoteTargetFilePath() const;
+
+    void onCheckIfAllowedToRenameComplete(const QVariantMap &values, QNetworkReply *reply = nullptr);
```
Comment (allexzander) [inline src/libsync/caseclashconflictsolver.h:70]: are we not supposed to make all `void` methods `slots`?

File: src/libsync/caseclashconflictsolver.h [hunk 9]
```diff
@@ -0,0 +1,95 @@
+#ifndef CASECLASHCONFLICTSOLVER_H
+#define CASECLASHCONFLICTSOLVER_H
```
Comment (allexzander) [inline src/libsync/caseclashconflictsolver.h:2]: why not use `#pragma once` as elsewhere?

File: src/libsync/filesystem.cpp [hunk 10]
```diff
@@ -128,6 +131,12 @@ qint64 FileSystem::getSize(const QString &filename)
         // Qt handles .lnk as symlink... https://doc.qt.io/qt-5/qfileinfo.html#details
         return getSizeWithCsync(filename);
     }
+#endif
+#ifdef QT_DEBUG
+    if (!QFileInfo::exists(filename)) {
```
Comment (allexzander) [inline src/libsync/filesystem.cpp:136]: is this just a leftover from debugging?

File: src/libsync/filesystem.cpp [hunk 11]
```diff
@@ -93,11 +94,13 @@ bool FileSystem::fileChanged(const QString &fileName,
...
 
 bool FileSystem::verifyFileUnchanged(const QString &fileName,
-    qint64 previousSize,
-    time_t previousMtime)
+                                     qint64 previousSize,
+                                     time_t previousMtime)
 {
-    const qint64 actualSize = getSize(fileName);
-    const time_t actualMtime = getModTime(fileName);
+    Q_ASSERT(QFileInfo::exists(fileName));
```
Comment (allexzander) [inline src/libsync/filesystem.cpp:100]: won't work in Release

File: src/libsync/propagatedownload.cpp [hunk 12]
```diff
@@ -1199,15 +1201,28 @@ void PropagateDownloadFile::downloadFinished()
     // Apply the remote permissions
     FileSystem::setFileReadOnlyWeak(_tmpFile.fileName(), !_item->_remotePerm.isNull() && !_item->_remotePerm.hasPermission(RemotePermissions::CanWrite));
 
-    bool isConflict = _item->_instruction == CSYNC_INSTRUCTION_CONFLICT
-        && (QFileInfo(fn).isDir() || !FileSystem::fileEquals(fn, _tmpFile.fileName()));
+    const bool isConflict = (_item->_instruction == CSYNC_INSTRUCTION_CONFLICT
```
Comment (allexzander) [inline src/libsync/propagatedownload.cpp:1204]: `const auto isConflict`

File: src/libsync/syncengine.cpp [hunk 13]
```diff
@@ -317,6 +317,20 @@ void SyncEngine::conflictRecordMaintenance()
...
 }
 
+void SyncEngine::caseClashConflictRecordMaintenance()
+{
+    // Remove stale conflict entries from the database
+    // by checking which files still exist and removing the
+    // missing ones.
+    const auto conflictRecordPaths = _journal->caseClashConflictRecordPaths();
+    for (const auto &path : conflictRecordPaths) {
+        auto fsPath = _propagator->fullLocalPath(QString::fromUtf8(path));
```
Comment (allexzander) [inline src/libsync/syncengine.cpp:327]: `const auto fsPath`

File: CMakeLists.txt (hunk 4 above)
Comment (mgallien) [reply CMakeLists.txt:172]: to get proper usage of the option within cmake build system
like for example being able to turn it on/off from Qt Creator

File: src/libsync/owncloudpropagator.cpp [hunk 14]
```diff
@@ -913,6 +913,63 @@ bool OwncloudPropagator::createConflict(const SyncFileItemPtr &item,
     return true;
 }
 
+OCC::Optional<QString> OwncloudPropagator::createCaseClashConflict(const SyncFileItemPtr &item, const QString &temporaryDownloadedFile)
+{
+    auto filename = QString{};
```
[the remark this answers is not in the document]
Comment (mgallien) [reply src/libsync/owncloudpropagator.cpp:918]: I disagree
I do not see a good reason not to use it
I have always find that almost always auto is a good rule
we should have auto everywhere unless there is a reason it cannot be used

File: src/libsync/owncloudpropagator.cpp [hunk 15]
```diff
@@ -1473,4 +1530,23 @@ QString OwncloudPropagator::remotePath() const
...
 
+void PropagateIgnoreJob::start()
+{
+    SyncFileItem::Status status = _item->_status;
+    if (status == SyncFileItem::NoStatus) {
+        if (_item->_instruction == CSYNC_INSTRUCTION_ERROR) {
+            status = SyncFileItem::NormalError;
+        } else {
+            status = SyncFileItem::FileIgnored;
+            ASSERT(_item->_instruction == CSYNC_INSTRUCTION_IGNORE);
```
[the remark this answers is not in the document]
Comment (mgallien) [reply src/libsync/owncloudpropagator.cpp:1541]: the `ASSERT` macro will also print a message in release mode if it fails

File: src/gui/caseclashfilenamedialog.h (hunk 7 above)
Comment (mgallien) [reply src/gui/caseclashfilenamedialog.h]: I improved the copyright notices

File: src/libsync/caseclashconflictsolver.h (hunk 9 above)
Comment (mgallien) [reply src/libsync/caseclashconflictsolver.h]: this is now done with `#pragma once`

**Review by allexzander — APPROVED** [review]

**Review by claucambra — APPROVED** [review]
Please make sure to remove the merge commit before merging this PR

**Comment by mgallien** [comment]
> Merge commit probably needs to be removed before merging this PR, otherwise looks good

done
thanks for the notice
